### PR TITLE
Support "max_pages" and "max_items" on "pages" generator

### PR DESCRIPTION
**Pull request by era**
Hello :),

I saw the issue https://github.com/vintasoftware/tapioca-wrapper/issues/28 and I tryied to implement the change. Could you verify if this was what you were thinking? And if not, could you please help me to make the necessary changes?

Cheers,

### CODE REVIEW

**Comment by filipeximenes** [comment]
Great! Thanks very much!
You got the ideia correctly, I just have comments about the code.

File: tapioca/tapioca.py [hunk 1]
```diff
@@ -170,14 +170,24 @@ def _get_iterator_next_request_kwargs(self):
         return self._api.get_iterator_next_request_kwargs(
             self._request_kwargs, self._data, self._response)
 
+    def _has_next(self, actual, kwargs):
+        has_more_page = "max_pages" not in kwargs or kwargs["max_pages"] > actual["page"]
+        has_more_items = "max_items" not in kwargs or kwargs["max_items"] > actual["items"]
+        return  has_more_page and has_more_items
+
     def pages(self, **kwargs):
```
Comment (filipeximenes) [inline tapioca/tapioca.py]: Explicitly pass the two new kwords in the method definition:
``def pages(self, max_pages=None, max_items=None,  **kwargs):``

File: tapioca/tapioca.py [hunk 2]
```diff
@@ -170,14 +170,24 @@ def _get_iterator_next_request_kwargs(self):
...
         executor = self
         iterator_list = executor._get_iterator_list()
+        actual = {"page": 0, "items": 0}
```
Comment (filipeximenes) [inline tapioca/tapioca.py]: No need for dictionary here, you can define two separate variable for this: current_page, current_item

**Comment by era** [comment]
Thank you for the feedback, I just changed the code. Could you please verify?

File: tapioca/tapioca.py [hunk 3]
```diff
@@ -170,14 +170,24 @@ def _get_iterator_next_request_kwargs(self):
...
 
-        while iterator_list:
+        while iterator_list and self._has_next(actual, kwargs):
             for item in iterator_list:
                 yield self._wrap_in_tapioca(item)
-
+                actual["items"] += 1
+                if not self._has_next(actual, kwargs):
```
Comment (filipeximenes) [inline tapioca/tapioca.py]: it's more redable if you make the item verification right here:
``if max_items and item_count < max_items:``

File: tapioca/tapioca.py [hunk 4]
```diff
@@ -170,14 +170,24 @@ def _get_iterator_next_request_kwargs(self):
...
+                    break
+                    
+            actual["page"] += 1
```
Comment (filipeximenes) [inline tapioca/tapioca.py]: it's more redable if you make the page_count verification here

File: tapioca/tapioca.py [hunk 5]
```diff
@@ -170,14 +170,24 @@ def _get_iterator_next_request_kwargs(self):
...
+        has_more_items = "max_items" not in kwargs or kwargs["max_items"] > actual["items"]
+        return  has_more_page and has_more_items
+
     def pages(self, **kwargs):
         executor = self
         iterator_list = executor._get_iterator_list()
+        actual = {"page": 0, "items": 0}
 
-        while iterator_list:
+        while iterator_list and self._has_next(actual, kwargs):
```
Comment (filipeximenes) [inline tapioca/tapioca.py]: No need for verifying ``_has_next`` on the loop definition.

**Comment by filipeximenes** [comment]
I made some other comments, can you also review them? 
item_count and page_count are better names for the variables, sorry for misleading you the first time

File: tapioca/tapioca.py (hunk 4 above)
Comment (era) [inline tapioca/tapioca.py]: Outside the for loop and before next_request_kwargs = executor._get_iterator_next_request_kwargs()?

Comment (filipeximenes) [inline tapioca/tapioca.py]: yes, just after you update the page_count

File: tapioca/tapioca.py (hunk 3 above)
Comment (era) [inline tapioca/tapioca.py]: If I understood the code correctly, the while loop iterates in the pages and the for in the items of that page, right?
If I made the verification without the function I will have to repeat it outside the for loop (to break the while loop avoiding new requests to the server). Should I proceed that way?

Comment (filipeximenes) [inline tapioca/tapioca.py]: You are right, my mistake.
Keep the method, just change it's name to something like ``_reached_max_limits`` to make it more clear.

**Comment by filipeximenes** [comment]
One more thing: add yourself to the contributors list :)

**Comment by era** [comment]
Done :), thank you so much for the time spent helping me :).

If you see something else that needs change, please let me know.

File: tapioca/tapioca.py [hunk 6]
```diff
@@ -170,13 +170,27 @@ def _get_iterator_next_request_kwargs(self):
...
 
         while iterator_list:
             for item in iterator_list:
                 yield self._wrap_in_tapioca(item)
+                item_count += 1
+                if self._reached_max_limits(page_count, item_count,  max_pages, max_items):
+                    break
+
+            page_count += 1
+            if self._reached_max_limits(page_count, item_count,  max_pages, max_items):
```
Comment (filipeximenes) [inline tapioca/tapioca.py]: Better move this to the beginning of the while. This way we will cover max_pages=0. Just for the sake of correctness.

File: tapioca/tapioca.py [hunk 7]
```diff
@@ -170,13 +170,27 @@ def _get_iterator_next_request_kwargs(self):
...
         executor = self
         iterator_list = executor._get_iterator_list()
+        page_count = 0
+        item_count = 0
 
         while iterator_list:
             for item in iterator_list:
                 yield self._wrap_in_tapioca(item)
+                item_count += 1
+                if self._reached_max_limits(page_count, item_count,  max_pages, max_items):
```
Comment (filipeximenes) [inline tapioca/tapioca.py]: Same here, move to the beginning of the for just to make sure we cover max_items=0.

Comment (filipeximenes) [inline tapioca/tapioca.py]: testing for ``max_items=0`` and ``max_pages=0`` would also be nice.

**Comment by era** [comment]
I changed the "if" position and added the test with zero :)

**Comment by filipeximenes** [comment]
Awesome! Thanks a lot!

**Pull request merged**